synapse/federation/sender/__init__.py

```diff
@@ -434,7 +434,23 @@ async def handle_event(event: EventBase) -> None:
                         # If there are no prev event IDs then the state is empty
                         # and so no remote servers in the room
                         destinations = set()
-                    else:
+
+                    if destinations is None:
+                        # During partial join we use the set of servers that we got
+                        # when beginning the join. It's still possible that we send
+                        # events to servers that left the room in the meantime, but
+                        # we consider that an acceptable risk since it is only our own
+                        # events that we leak and not other server's ones.
+                        partial_state_destinations = (
+                            await self.store.get_partial_state_servers_at_join(
+                                event.room_id
+                            )
+                        )
+
+                        if len(partial_state_destinations) > 0:
+                            destinations = partial_state_destinations
```

Comment:
I just realised that we won't send events to servers that join after us.

`self._storage_controllers.state.get_current_hosts_in_room_or_partial_state_approximation(room_id)` returns a list that does include those servers (apologies for the naming), but it never returns an empty list.
We may want to go back to the original structure (`elif await self.store.is_partial_state_room(event.room_id):`) but use that method instead.

Sorry for not catching this earlier!

Reply:
I suppose there's a question here: how much are we happy to trust this approximation? In prior discussions I gathered that we were just going to stick with the original list handed out over `/send_join` — is that no longer the case?
I guess the client will see the memberships for the 'extras', even if we end up rejecting them later on, it won't ever send events to servers that the user couldn't see, so maybe it's OK?

Reply:
I'm missing / have forgotten context from prior discussions. If we've decided to go with only the original list, then please ignore the comment!

It's hard to say what memberships an end user might actually see, given that they'll be doing lazy loading syncs, which only provide memberships when a user sends a message. I think the member list in Element Web might not work until the room is fully synced.

Reply:
I haven't merged that yet because I am unsure.

To my understanding, the approximation is initial list + updates from membership events that we received after.
There is a possibility that we can't validate those events out of missing state, hence I think we probably want to not use the approximation ?
We also can't be sure of the initial `servers_in_room` but at least we trust only one server there.

In the grand scheme of things it's not a big trouble because we potentially only leak our own events.

For now I am leaning towards merging it like that.